src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/users/RestUserGenerator.kt

```diff
@@ -48,17 +61,38 @@ class RestUserGenerator(readTimeout: Duration) : UserGenerator {
                 throw e
             }
         }.use { response ->
-            if (response.code() == 201) {
+            if (response.code() == HttpURLConnection.HTTP_CREATED) {
                 logger.info("Created a new user $userName")
+                return userName
             } else {
-                val msg = "Failed to create a new user $userName:" +
+                val possibleReason = if (response.code() == HttpURLConnection.HTTP_BAD_REQUEST) {
+                    givenNames.compareAndSet(givenNamesUsedInThisAttempt, givenNamesUsedInThisAttempt+1)
```

Comment:
When we lose the CAS (returns `false`), we lose the increment by design?
TBH I'm a bit lost with these counters and the meaning of `givenNames` in the `nameGenerator.pickRandomUnique(givenNames)` call.

Reply:
Yes. We want to lose the increment because it means somebody else incremented it too, perhaps more than once.
I've renamed givenNames to givenNamesCnt now.

Reply:
I was confused about behavior. It was clear it's a counter.
`nameGenerator.pickRandomUnique(3)` looks like it should generate 3 names.

So I looked into the source code. And it seems it's about having multiple given names, like "Adam Ben Charlie $surname".

So coming back to the counters and increments, we keep increasing the number of names in order to decrease likelihood of collisions?

It's a bit mind-bending. Ideally we should keep `RestUserGenerator` focused on REST and keep randomization and collision avoidance in a separate unit. If the collision detection is impossible to untangle from REST (due to HTTP response codes?), then let's use var names, which explain the purpose. E.g. `givenNames` could be `nameCollisions`.

Reply:
I've extracted the code to a separate class, it will at least reduce the amount of globals in the REST code. I don't think we can decouple it like you want it to, we could only make it slightly simpler if we assume that RestUserGenerator will never be called by multiple threads. Then we could remove the part that stores the current given name counter and keep only the pick() and increase() calls.

Reply:
I see.
[`UserGenerator`s should be thread-safe](https://github.com/atlassian/virtual-users/blob/release-3.10.0/src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/users/UserGenerator.kt#L12-L13), so we have to pay the concurrency complexity.
Thanks for simplifying RUG a bit.